File: python/purehtml/transformers/transformer_factory.py
import re
from typing import List

from purehtml.transformers.transformer import Transformer
from purehtml.transformers.transformers import Transformers


class TransformerFactory:
    # Regular expression to match the transformer name
    TRANSFORMER_NAME_REGEX = re.compile(r"^([\w-]+[^(])")

    @staticmethod
    def create(transform: str) -> 'Transformer':
        # Extract the transformer name and arguments
        transformer_name = TransformerFactory.extract_transformer_name(transform)
        args = TransformerFactory.extract_transformer_args(transform)

        # Get the transformer instance by name
        transformer_impl = Transformers.get_by_name(transformer_name)

        if transformer_impl is None:
            raise ValueError(f"Transformer with name \"{transformer_name}\" not found.")

        # Set arguments for the transformer
        transformer_impl.set_args(args)
        return transformer_impl

    @staticmethod
    def extract_transformer_name(definition: str) -> str:
        """
        Extracts the transformer name from the definition string.
        :param definition: The full transformer definition (name + arguments).
        :return: The transformer name as a string.
        """

        matcher = TransformerFactory.TRANSFORMER_NAME_REGEX.match(definition)
        if not matcher:
            raise ValueError(f"Invalid transformer name: {definition}")
        return matcher.group(0)

    @staticmethod
    def extract_transformer_args(definition: str) -> List[str]:
        """
        Extracts the transformer arguments from the definition string.
        :param definition: The full transformer definition (name + arguments).
        :return: A list of arguments (strings).
        """
        open_index = definition.find("(")
        if open_index == -1:
            return []

        close_index = definition.rfind(")")
        if close_index == -1 or close_index < open_index:
            return []

        return TransformerFactory._parse_arg_list(definition[open_index + 1:close_index])

    @staticmethod
    def _parse_arg_list(args_str: str) -> List[str]:
        """
        Parses a comma-separated argument list, honoring single- and
        double-quoted arguments so that delimiters, regex patterns, and
        replacement strings can contain commas, spaces, and other
        characters that would otherwise conflict with the comma separator,
        e.g. replace("\\s+", " ") or split(", "). Inside a quoted argument,
        a backslash escapes the matching quote character or another
        backslash (\\" or \\\\); any other backslash (e.g. the \\s above)
        is left untouched so regex patterns pass through unchanged. Bare
        (unquoted) arguments are trimmed of surrounding whitespace but
        otherwise taken verbatim.
        """
        args: List[str] = []
        i = 0
        n = len(args_str)

        while i < n:
            while i < n and args_str[i].isspace():
                i += 1
            if i >= n:
                break

            ch = args_str[i]
            if ch in ("\"", "'"):
                quote = ch
                i += 1
                value_chars = []
                while i < n and args_str[i] != quote:
                    if args_str[i] == "\\" and i + 1 < n and args_str[i + 1] in (quote, "\\"):
                        value_chars.append(args_str[i + 1])
                        i += 2
                    else:
                        value_chars.append(args_str[i])
                        i += 1
                i += 1  # skip closing quote
                args.append("".join(value_chars))
            else:
                start = i
                while i < n and args_str[i] != ",":
                    i += 1
                args.append(args_str[start:i].strip())

            while i < n and args_str[i].isspace():
                i += 1
            if i < n and args_str[i] == ",":
                i += 1

        return args

File: java/src/main/java/com/purescraps/purehtml/transformers/TransformerFactory.java
package com.purescraps.purehtml.transformers;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class TransformerFactory {

    private static final Pattern TRANSFORMER_NAME_REGEX = Pattern.compile("^([\\w-]+[^(])");

    public static Transformer create(String transform) {

        String transformerName = extractTransformerName(transform);
        List<String> args = extractTransformerArgs(transform);

        Transformer transformerImpl = Transformers.getByName(transformerName);
        if (transformerImpl == null) {
            throw new IllegalArgumentException("Transformer with name \"" + transformerName + "\" not found.");
        }
        transformerImpl.setArgs(args);
        return transformerImpl;

    }

    private static String extractTransformerName(String def) {
        Matcher matcher = TRANSFORMER_NAME_REGEX.matcher(def);
        if (!matcher.find()) {
            throw new IllegalArgumentException("Invalid transformer name: " + def);
        }
        return matcher.group(0);
    }

    private static List<String> extractTransformerArgs(String def) {
        int openIndex = def.indexOf('(');
        if (openIndex == -1) {
            return new ArrayList<>();
        }

        int closeIndex = def.lastIndexOf(')');
        if (closeIndex == -1 || closeIndex < openIndex) {
            return new ArrayList<>();
        }

        return parseArgList(def.substring(openIndex + 1, closeIndex));
    }

    /**
     * Parses a comma-separated argument list, honoring single- and
     * double-quoted arguments so that delimiters, regex patterns, and
     * replacement strings can contain commas, spaces, and other characters
     * that would otherwise conflict with the comma separator, e.g.
     * replace("\s+", " ") or split(", "). Inside a quoted argument, a
     * backslash escapes the matching quote character or another backslash
     * (\" or \\); any other backslash (e.g. the \s above) is left untouched
     * so regex patterns pass through unchanged. Bare (unquoted) arguments
     * are trimmed of surrounding whitespace but otherwise taken verbatim.
     */
    private static List<String> parseArgList(String argsStr) {
        List<String> args = new ArrayList<>();
        int i = 0;
        int n = argsStr.length();

        while (i < n) {
            while (i < n && Character.isWhitespace(argsStr.charAt(i))) {
                i++;
            }
            if (i >= n) {
                break;
            }

            char ch = argsStr.charAt(i);
            if (ch == '"' || ch == '\'') {
                char quote = ch;
                i++;
                StringBuilder value = new StringBuilder();
                while (i < n && argsStr.charAt(i) != quote) {
                    char c = argsStr.charAt(i);
                    char next = (i + 1 < n) ? argsStr.charAt(i + 1) : '\0';
                    if (c == '\\' && i + 1 < n && (next == quote || next == '\\')) {
                        value.append(next);
                        i += 2;
                    } else {
                        value.append(c);
                        i++;
                    }
                }
                i++; // skip closing quote
                args.add(value.toString());
            } else {
                int start = i;
                while (i < n && argsStr.charAt(i) != ',') {
                    i++;
                }
                args.add(argsStr.substring(start, i).trim());
            }

            while (i < n && Character.isWhitespace(argsStr.charAt(i))) {
                i++;
            }
            if (i < n && argsStr.charAt(i) == ',') {
                i++;
            }
        }

        return args;
    }
}
